[Samuel Lee] iP

sam-theman88 opened the pull request:
# Edith

> "Failure is only the opportunity to begin again. Only this time, more wisely." - Uncle Iroh, Avatar The Last Airbender ([source](https://www.goodreads.com/author/quotes/12618081.Uncle_Iroh))

Free yourself from needlessly worrying about whether you've remembered to finish a task. With Edith, it's
- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,
1. download it from this repo
2. double-click it
3. add your tasks
4. let it manage your tasks for you 🤩 

And the best part? It is completely **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [x] Managing events
- [ ] Reminders (coming soon)

***

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```ruby
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

sam-theman88 committed:
Renamed chatbot to Edith, and added greeting

sam-theman88 committed:
Implemented initial skeletal version

sam-theman88 committed:
Implemented OOP paradigm

sam-theman88 committed:
Chatbot can now receive user input, and added echo feature

sam-theman88 committed:
Updated the echo feature into an add feature

sam-theman88 committed:
Added list feature to display all added tasks when requested

sam-theman88 committed:
Abstracted a task into Task class, and updated add feature accordingly

sam-theman88 committed:
Added mark and unmark features

sam-theman88 committed:
Updated mark and unmark features to display proper output to users

sam-theman88 committed:
Fixed indexing bug for the mark and unmark features

sam-theman88 committed:
Updated display output of add feature

sam-theman88 committed:
Added 3 types of tasks as subclasses of Task class

sam-theman88 committed:
Updated add and list features to account for the 3 types of tasks

sam-theman88 committed:
Semi-automated the testing of Edith, and ran the testing

sam-theman88 committed:
Edited a method name to markTaskUndone()

sam-theman88 committed:
Created EdithException class, and updated Edith to handle incorrect inputs from user

sam-theman88 committed:
Added delete feature to Edith

sam-theman88 committed:
Add a save feature to Edith

sam-theman88 committed:
Merge branch 'branch-Level-7' to branch 'master'

sam-theman88 committed:
Update Edith to understand dates and times properly and not as a String

sam-theman88 committed:
Merge branch 'branch-Level-8' to branch 'master'

sam-theman88 committed:
Abstract out the user interaction aspect of Edith into a class

sam-theman88 committed:
Add a method to show a specified error message

sam-theman88 committed:
Abstract out the storage functionality of Edith into a class

sam-theman88 committed:
Abstract out the parsing functionality of user input into a class

sam-theman88 committed:
Update Task into an abstract class

sam-theman88 committed:
Abstract out the task list and its relevant operations into a class

sam-theman88 committed:
Abstract out the giving of commands to Edith into an abstract class

sam-theman88 committed:
Add a getter to retrieve the list of tasks

sam-theman88 committed:
Abstract out the commands for Edith into child classes of Command

sam-theman88 committed:
Update class after abstracting out various aspects of the chatbot

sam-theman88 committed:
Merge branch 'branch-A-MoreOOP' to branch 'master'

sam-theman88 committed:
Organise the classes into their relevant packages

sam-theman88 committed:
Organise the classes into their relevant packages

sam-theman88 committed:
Organise the classes into their relevant packages

sam-theman88 committed:
Merge branch 'branch-A-Packages' to branch 'master'

sam-theman88 committed:
Merge remote-tracking branch 'origin/add-gradle-support' to branch 'master'

sam-theman88 committed:
Add gradle support to build and run edith

sam-theman88 committed:
Add text file for save and load feature

sam-theman88 committed:
Add JUnit tests to some non-trivial public methods

sam-theman88 committed:
Merge branch 'branch-A-JUnit' to branch 'master'

sam-theman88 committed:
Add JavaDoc comments to the code

sam-theman88 committed:
Merge branch 'branch-A-JavaDoc' to branch 'master'

sam-theman88 committed:
Add JavaDoc comments to the code

sam-theman88 committed:
Resolve the conflict when merging branch-A-CodingStandard to master

sam-theman88 committed:
Add the find feature to edith

sam-theman88 committed:
Resolve the conflict when merging branch-Level-9 to master

xGladiate commented on src/main/java/edith/Storage.java:
```diff
@@ -0,0 +1,111 @@
+     * @throws IOException If an I/O error occurs while reading from the file.
+     * @throws EdithException If the file contains corrupted data or the data format is invalid.
+     */
+    public ArrayList<Task> load() throws IOException, EdithException {
```
Might want use a more specific method name, eg. `loadTasks()`.

xGladiate commented on src/test/java/edith/command/ListOnDateCommandTest.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    @Test
+    public void execute_validDateWithTasks_success() {
```
Good job for following the naming conventions for writing test methods! 💯

xGladiate commented on src/main/java/edith/Parser.java:
Good job, all classes inheriting from the `command` class could be easily understood. 
However, you might want to consider an alternative to the class `ListOnDateCommand` which might be a bit hard to decipher what the class does based on its name.

xGladiate commented on src/main/java/edith/Ui.java:
Good job for making the method name clear through the use of 'show' prefix! 💯

xGladiate reviewed (commented):
Overall clear and concise code quality, keep up the good work!!💯

sam-theman88 committed:
Implement checkStyle to detect coding style violations

sam-theman88 committed:
Merge branch 'branch-A-CheckStyle' to branch 'master'

sam-theman88 committed:
Add relevant classes and files to update edith to a GUI app

sam-theman88 committed:
Update commands for GUI

sam-theman88 committed:
Update relevant methods for GUI

sam-theman88 committed:
Update edith to a GUI app

sam-theman88 committed:
Add JavaDocs for classes added to make the GUI

sam-theman88 committed:
Merge branch 'branch-Level-10' to branch 'master'

sam-theman88 committed:
Add exit logic to close GUI app

sam-theman88 committed:
Add assertions to enforce valid state assumptions in Task class

There are some key assumptions in the Task class. Adding Java assertions
in the class will document and enforce these assumptions that should
hold.

For example, the task's description should not be null or empty when
creating a new task.

These assertions help ensure that the task object is always in a valid
state during its lifecycle.

sam-theman88 committed:
Add assertions to validate task list operations in TaskList class

There are some key assumptions in the TaskList class. Adding Java
assertions in the class will document and enforce these assumptions
that should hold.

For example, the task index should be within valid bounds when removing
or retrieving tasks from the list, and only non-null tasks should be
added to the task list.

These assertions serve as a safety net to catch potential errors during
task list management. Hence doing so will improve its reliability.

sam-theman88 committed:
Apply refactoring to Storage class to improve code readability

sam-theman88 committed:
Apply refactoring to AddCommand class to improve code readability

sam-theman88 committed:
Fix indexing bug in error message

sam-theman88 committed:
Apply refactoring to TaskList class to improve code readability

sam-theman88 committed:
Merge pull request #2 from sam-theman88/branch-A-Assertions

Merge branch 'branch-A-Assertions' to branch 'master'

sam-theman88 committed:
Merge branch 'master' to branch 'branch-A-CodeQuality'

sam-theman88 committed:
Merge pull request #3 from sam-theman88/branch-A-CodeQuality

Merge branch 'branch-A-CodeQuality' to branch 'master'

sam-theman88 committed:
Use GitHub Actions to set up Continuous Integration

sam-theman88 committed:
Merge branch 'branch-A-CI' to branch 'master'

sam-theman88 committed:
Delete .github/workflows directory

sam-theman88 committed:
Delete config directory

sam-theman88 committed:
Update gradle build plugins

sam-theman88 committed:
Make the command syntax of Edith more flexible for users

sam-theman88 committed:
Merge branch 'branch-C-FriendlierSyntax' to branch 'master'

sam-theman88 committed:
Update edith to greet users when first opened

sam-theman88 committed:
Update edith to greet users bye before closing the app

sam-theman88 committed:
Update JUnit tests for some non-trivial public methods

sam-theman88 committed:
Update the exit logic for the app

sam-theman88 committed:
Update gui elements to automatically resize

sam-theman88 committed:
Improve the ui and ux of the gui

sam-theman88 committed:
Merge branch 'branch-A-BetterGui' to branch 'master'

sam-theman88 committed:
Update gui element for error messages

sam-theman88 committed:
Update JavaDocs for respective classes and methods

sam-theman88 committed:
Update min width for gui window

sam-theman88 committed:
Add a representative screenshot of the chatbot product

sam-theman88 committed:
Update README to be user guide for app

sam-theman88 committed:
Update the README doc

sam-theman88 committed:
Update the README doc

sam-theman88 committed:
Update the README doc
